sample/src/main/java/jp/co/cyberagent/android/gpuimage/sample/utils/Camera2Loader.kt

```diff
@@ -113,23 +118,26 @@ class Camera2Loader(val activity: Activity) : CameraLoader() {
     }
 
     private fun chooseOptimalSize(): Size {
-        val backCameraId = getCameraId(CameraCharacteristics.LENS_FACING_BACK) ?: return Size(0, 0)
-        val frontCameraId =
-            getCameraId(CameraCharacteristics.LENS_FACING_FRONT) ?: return Size(0, 0)
-        val backSizes =
-            cameraManager.getCameraCharacteristics(backCameraId)
-                .get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP)
-                ?.getOutputSizes(ImageFormat.YUV_420_888)
-        val frontSizes =
-            cameraManager.getCameraCharacteristics(frontCameraId)
-                .get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP)
-                ?.getOutputSizes(ImageFormat.YUV_420_888)
-
-        return backSizes?.filter {
-            frontSizes?.contains(it) ?: true
-        }?.maxBy {
-            it.width * it.height
-        } ?: Size(0, 0)
+        if (viewWidth == 0 || viewHeight == 0) {
+            return Size(0, 0)
+        }
+        val cameraId = getCameraId(cameraFacing) ?: return Size(0, 0)
+        val outputSizes = cameraManager.getCameraCharacteristics(cameraId)
+            .get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP)
+            ?.getOutputSizes(ImageFormat.YUV_420_888)
+        val optimalSize = outputSizes?.filter {
+            it.width * it.height > PREVIEW_WIDTH * PREVIEW_HEIGHT
+        }?.minBy {
+            (it.width / it.height - viewWidth / viewHeight).absoluteValue
+        }
+
+        return if (optimalSize == null) {
+            Size(PREVIEW_WIDTH, PREVIEW_HEIGHT)
+        } else if (optimalSize.width > viewWidth || optimalSize.height > viewHeight) {
+            Size(optimalSize.width / 2, optimalSize.height / 2)
```

Comment:
If calculated here, if outputSizes does not include the size? If so, it should be calculated during filtering.